va/wayland/va_wayland_linux_dmabuf.c

```diff
@@ -0,0 +1,364 @@
+    bool                      default_feedback_done;
+} VADisplayContextWaylandLinuxDmabuf;
+
+static void
+feedback_handle_done(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback
+)
+{
+    VADisplayContextP const pDisplayContext = data;
+    struct va_wayland_linux_dmabuf_context *wl_linux_dmabuf_ctx = pDisplayContext->opaque;
+
+    wl_linux_dmabuf_ctx->default_feedback_done = true;
+
+    zwp_linux_dmabuf_feedback_v1_destroy(feedback);
+}
+
+static void
+feedback_handle_format_table(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback,
+    int                                  fd,
+    uint32_t                             size
+)
+{
+    close(fd);
+}
+
+/* XXX: replace with drmGetDeviceFromDevId() */
+static drmDevice *
+get_drm_device_from_dev_id(dev_t dev_id)
+{
+    uint32_t flags = 0;
+    int devices_len, i, node_type;
+    drmDevice *match = NULL, *dev;
+    struct stat statbuf;
+
+    devices_len = drmGetDevices2(flags, NULL, 0);
+    if (devices_len < 0) {
+        return NULL;
+    }
+    drmDevice **devices = calloc(devices_len, sizeof(*devices));
+    if (devices == NULL) {
+        return NULL;
+    }
+    devices_len = drmGetDevices2(flags, devices, devices_len);
+    if (devices_len < 0) {
+        free(devices);
+        return NULL;
+    }
+
+    for (i = 0; i < devices_len; i++) {
+        dev = devices[i];
+        for (node_type = 0; node_type < DRM_NODE_MAX; node_type++) {
+            if (!(dev->available_nodes & (1 << node_type)))
+                continue;
+
+            if (stat(dev->nodes[node_type], &statbuf) != 0) {
+                va_wayland_error("stat() failed for %s", dev->nodes[node_type]);
+                continue;
+            }
+
+            if (statbuf.st_rdev == dev_id) {
+                match = dev;
+                break;
+            }
+        }
+    }
+
+    for (i = 0; i < devices_len; i++) {
+        dev = devices[i];
+        if (dev != match)
+            drmFreeDevice(&dev);
+    }
+    free(devices);
+
+    return match;
+}
+
+static void
+feedback_handle_main_device(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback,
+    struct wl_array                     *device_array
+)
+{
+    dev_t dev_id;
+    drmDevice *dev;
+    const char *dev_path;
+    VADisplayContextP const pDisplayContext = data;
+    VADriverContextP const ctx = pDisplayContext->pDriverContext;
+    struct drm_state * const drm_state = ctx->drm_state;
+
+    assert(device_array->size == sizeof(dev_id));
+    memcpy(&dev_id, device_array->data, sizeof(dev_id));
+
+    dev = get_drm_device_from_dev_id(dev_id);
+    if (!dev) {
+        va_wayland_error("failed to get DRM device from device ID");
+        return;
+    }
+
+    if (!(dev->available_nodes & (1 << DRM_NODE_RENDER)))
+        goto end;
+
+    dev_path = dev->nodes[DRM_NODE_RENDER];
```

Comment:
always render node? will not touch drm master node

Reply:
Yeah, DRM master nodes are for KMS only nowadays.